src/test/blake2s_test.cpp

```diff
@@ -0,0 +1,234 @@
+#include "circuits/blake2s/blake2s_comp.hpp"
+#include "circuits/blake2s/g_primitive.hpp"
+#include "snarks_alias.hpp"
+
+#include "gtest/gtest.h"
+#include <libff/common/default_types/ec_pp.hpp>
+
+// Access the `from_bits` function and other utils
+#include "circuits/circuits-util.hpp"
+#include "util.hpp"
+
+using namespace libsnark;
+using namespace libzeth;
+
+typedef libff::default_ec_pp ppT;
+typedef libff::Fr<ppT> FieldT;
+
+namespace
+{
+
+// This test corresponds to the first call of the g_primitive of blake2s(b"hello
+// world"). As blake2s first formats the input blocks in 32bit words in little
+// endian, the inputs of the first g_primitive are "lleh" and "ow o"
+// ("hello world" -> "hell" "o wo" "rld" plus padding-> "lleh" "ow o" "dlr" plus
+// padding)
+TEST(TestG, TestTrue)
+{
+    libsnark::protoboard<FieldT> pb;
+
+    libsnark::pb_variable<FieldT> ZERO;
+    ZERO.allocate(pb, "zero");
+    pb.val(ZERO) = FieldT::zero();
+
+    libsnark::pb_variable_array<FieldT> a = from_bits(
+        {
+            0, 1, 1, 0, 1, 0, 1, 1, // 6B
+            0, 0, 0, 0, 1, 0, 0, 0, // 08
+            1, 1, 1, 0, 0, 1, 1, 0, // E6
+            0, 1, 0, 0, 0, 1, 1, 1  // 47
+        },
+        ZERO);
+
+    libsnark::pb_variable_array<FieldT> b = from_bits(
+        {
+            0, 1, 0, 1, 0, 0, 0, 1, // 51
+            0, 0, 0, 0, 1, 1, 1, 0, // 0E
+            0, 1, 0, 1, 0, 0, 1, 0, // 52
+            0, 1, 1, 1, 1, 1, 1, 1  // 7F
+        },
+        ZERO);
+
+    libsnark::pb_variable_array<FieldT> c = from_bits(
+        {
+            0, 1, 1, 0, 1, 0, 1, 0, // 6A
+            0, 0, 0, 0, 1, 0, 0, 1, // 09
+            1, 1, 1, 0, 0, 1, 1, 0, // E6
+            0, 1, 1, 0, 0, 1, 1, 1  // 67
+        },
+        ZERO);
+
+    libsnark::pb_variable_array<FieldT> d = from_bits(
+        {
+            0, 1, 0, 1, 0, 0, 0, 1, // 51
+            0, 0, 0, 0, 1, 1, 1, 0, // 0E
+            0, 1, 0, 1, 0, 0, 1, 0, // 52
+            0, 1, 1, 1, 0, 1, 0, 0  // 74
+        },
+        ZERO);
+
+    // First word in little endian "lleh"
+    libsnark::pb_variable_array<FieldT> x = from_bits(
+        {
+            0, 1, 1, 0, 1, 1, 0, 0, // 6C
+            0, 1, 1, 0, 1, 1, 0, 0, // 6C
+            0, 1, 1, 0, 0, 1, 0, 1, // 65
+            0, 1, 1, 0, 1, 0, 0, 0  // 68
+        },
+        ZERO);
+
+    // Second word in little endian "ow o"
+    libsnark::pb_variable_array<FieldT> y = from_bits(
+        {
+            0, 1, 1, 0, 1, 1, 1, 1, // 6F
+            0, 1, 1, 1, 0, 1, 1, 1, // 77
+            0, 0, 1, 0, 0, 0, 0, 0, // 20
+            0, 1, 1, 0, 1, 1, 1, 1  // 6F
+        },
+        ZERO);
```

Comment:
So, you do not pass the final block "dlr" to the hash function to retrieve the hash of "hello world"?

Reply:
This test tests g_primitive which only takes as inputs some state (a,b,c,d here) and _two_ words (x and y here). 

I made the choice to choose as test vector the first g_primitive of the first round of the compression of block input "hello world". In that case g_primitive takes as input only the first two words ("lleh" and "ow o".

If I want to test the final block "dlr", it means I am testing  g_primitive another time which I thought was not necessary.

If you want to add this test, I need to pad "dlr", define y as some padding, update the state (a,b,c,d) as the output of the first g_primitive gadget and compute the output.